cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go

```diff
@@ -0,0 +1,256 @@
+	ClientSetFromFile(string) (*clientset.Clientset, error)
+	InitCfg() (*kubeadmapi.InitConfiguration, error)
+}
+
+func getPhaseDescription(component string) string {
+	return fmt.Sprintf("Generates the %s static Pod manifest", component)
+}
+
+func getControlPlaneJoinPhaseFlags() []string {
+	return []string{
+		options.APIServerAdvertiseAddress,
+		options.APIServerBindPort,
+		options.CfgPath,
+		options.ControlPlane,
+		options.NodeName,
+		options.TokenDiscovery,
+		options.TokenDiscoveryCAHash,
+		options.TokenDiscoverySkipCAHash,
+	}
+}
+
+// NewControlPlaneJoinPhase creates a kubeadm workflow phase that implements marking the new node as master and update the cluster status
+// with information about current node
+func NewControlPlaneJoinPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "control-plane-join",
+		Short: "Prepares the machine to join a control plane.",
+		Long:  cmdutil.MacroCommandLongDescription,
+		Phases: []workflow.Phase{
+			{
+				Name:           "all",
+				Short:          "Generates all static Pod manifest files",
+				InheritFlags:   getControlPlaneJoinPhaseFlags(),
+				RunAllSiblings: true,
+			},
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeAPIServer),
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeControllerManager),
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeScheduler),
+			newEtcdLocalSubphase(),
+			newUploadConfigSubphase(),
+			newMarkControlPlaneSubphases(),
+		},
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newControlPlaneJoinSubphase(component string) workflow.Phase {
+	return workflow.Phase{
+		Name:         controlPlanePhaseProperties[component].name,
+		Short:        controlPlanePhaseProperties[component].short,
+		Run:          runControlPlaneJoinSubphase(component),
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newEtcdLocalSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:         "local",
+		Short:        "Generates the static Pod manifest file for a local etcd member",
+		Run:          runEtcdLocalPhase,
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newUploadConfigSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:         "upload-config",
+		Short:        "Upload the currently used configuration to the cluster",
+		Run:          runUploadConfigPhaseLocal,
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newMarkControlPlaneSubphases() workflow.Phase {
+	return workflow.Phase{
+		Name:         "mark-control-plane",
+		Short:        "Mark a node as a control-plane",
+		Run:          runMarkControlPlanePhaseLocal,
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func runEtcdLocalPhase(c workflow.RunData) error {
+	data, ok := c.(controlPlaneJoinData)
+	if !ok {
+		return errors.New("control-plane-join phase invoked with an invalid data struct")
+	}
+
+	kubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
```

Comment:
here we have a little complication.
the way this is done discards the `--kubeconfig` flag the user might want to pass and uses a hardcoded path.

please look at init.go:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/init.go#L400-L405

we need such a method for join.go

and also this added to join.go but for the `joinRunner`:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/init.go#L164-L165

and also this has to be added to the list of inherited flags as mentioned in a earlier comment:
`options.KubeconfigPath`
[1]

Reply:
Actually, it's even more complicated than that. The user may supply kubeconfig file in `data.Cfg().Discovery.File.KubeConfigPath` or via the `--discovery-file` command line switch. In that case we need to take those into account too.

@neolit123 @fabriziopandini WDYT?

Reply:
OTOH isn't this a 1:1 as how it was done before?

https://github.com/kubernetes/kubernetes/blob/f7c4389b793cd6cf0de8d67f2c5db28b3985ad59/cmd/kubeadm/app/cmd/join.go#L457

Reply:
@rosti `data.Cfg().Discovery.File.KubeConfigPath` allows to specify discovery info (master address, bootstrap-token and CA cert) using a kubeconfig file, but this should not mixed with the Kubeconfig file created during the control-plane prepare phase (that instead contains ClusterAdmin credentials). 
So is ok to apply the existing logic 1:1 as pointed by @ereslibre

Reply:
I was forgetting the addition of the `--kubeconfig` flag, please change this into
`kubeConfigFile := j.KubeConfigFile()`